site/content/blog/2021-04-23-optimizing-go-compiler-3.md

```diff
@@ -0,0 +1,234 @@
+
+The above rule is pretty much identical to the actual rule, which ended up looking like this:
+
+```
+(MOVBstore [7] p1 (SHRQconst [56] w)
+  x1:(MOVWstore [5] p1 (SHRQconst [40] w)
+  x2:(MOVLstore [1] p1 (SHRQconst [8] w)
+  x3:(MOVBstore p1 w mem))))
+  && x1.Uses == 1
+  && x2.Uses == 1
+  && x3.Uses == 1
+  && clobber(x1, x2, x3)
+  => (MOVQstore p1 w mem)
+```
+
+The only addition here is the `x1`, `x2` and `x3` labels, which name each of the values that correspond to the inner `MOV` instructions, and the boolean condition we have at the end of the matching pattern, which is actually literal Go code:
+
+```go
+x1.Uses == 1
+  && x2.Uses == 1
+  && x3.Uses == 1
+  && clobber(x1, x2, x3)
+```
+
+This boolean condition is simply an additional restriction to the pattern matching. The rules specification say that this pattern will only match if the pattern is matched **and** if this condition is true.
+
+For understanding what the additional restriction is, we need to understand what the `.Uses` field and the `clobber` function do with `x1`, `x2`, and `x3`, which are instances of [the `Value` struct](https://github.com/golang/go/blob/bf48163e8f2b604f3b9e83951e331cd11edd8495/src/cmd/compile/internal/ssa/value.go#L17-L63) defined in the compiler. This is actually pretty straightforward:
+
+-   The [`Uses` field](https://github.com/golang/go/blob/bf48163e8f2b604f3b9e83951e331cd11edd8495/src/cmd/compile/internal/ssa/value.go#L51-L52) is a field that we can access in every SSA value and, per its definition, is a counter of the number of times that specific value is used.
+-   The [`clobber` function](https://github.com/golang/go/blob/bf48163e8f2b604f3b9e83951e331cd11edd8495/src/cmd/compile/internal/ssa/rewrite.go#L935-L945) both checks that the value is never used again and decrements the `.Uses` counter.
+
+In short, what we are checking with the boolean condition is that each of the inner `MOV` instructions are used only once, and making sure that they are never used again. This is a simple strategy to restrict the pattern matching to the specific case we want to fix.
+
+# Wrapping up
+
+So... that's it! We did it!
+
+We wrote a failing test:
+
+```go
+func store_le64_load(b []byte, x *[8]byte) {
+	_ = b[8]
+	// amd64:-`MOV[BWL]`
+	binary.LittleEndian.PutUint64(b, binary.LittleEndian.Uint64(x[:]))
+}
+```
+
+And we added a new AMD64 rule:
+
+```
+(MOVBstore [7] p1 (SHRQconst [56] w)
+  x1:(MOVWstore [5] p1 (SHRQconst [40] w)
+  x2:(MOVLstore [1] p1 (SHRQconst [8] w)
+  x3:(MOVBstore p1 w mem))))
+  && x1.Uses == 1
+  && x2.Uses == 1
+  && x3.Uses == 1
+  && clobber(x1, x2, x3)
+  => (MOVQstore p1 w mem)
+```
+
+With the addition of this rule, the test passes, so we solved the issue!
+
+...or didn't we? Sorry, there's just one more wee thing to do. In the issue comments, [@mundaym](https://github.com/mundaym) pointed out [an important thing](https://github.com/golang/go/issues/41663#issuecomment-699904531):
```

Comment:
Should this be "or did we?"

Reply:
Fixed in https://github.com/mattermost/mattermost-developer-documentation/pull/803/commits/1c7164da0e23ad3fe203044b55e41bd39a29e1c7, thanks!